### ISSUE TITLE
Blender Addon for importing OpenMVG SfM reconstruction results

### ISSUE THREAD

**Issue by SBCV** [openMVG/openMVG#1585]
Hi Pierre,

I am working on a Blender Addon (https://github.com/SBCV/Blender-Addon-Photogrammetry-Importer) that allows you to import different photogrammetry data formats into Blender. The latest version supports now OpenMVG JSON files.

I can imagine that this tool could be useful for other OpenMVG JSON files users as well. It offers for example a nice way to visualize the reconstruction results (including cameras and image planes).

One can use Blender's camera animation tool to render the reconstruction. The camera animation in Blender offers many useful options to define the camera motion. You can define for example the camera path in 3D and add looking constraints for the camera.
Furthermore, you can load different models (with different file formats) into Blender at the same time. So you can for example render a point cloud and the corresponding mesh at the same time (e.g. to highlight differences).
I used it for example to compare reconstruction results of virtual data with the corresponding virtual environment.

For Blender users:
You can use OpenMVG instead of Blender's camera tracking tool to reconstruct the scene. Which is way more comfortable, since Blender's camera tracking requires a lot of user interaction to compute reasonable results.

Just wanted to inform you about that. Maybe you want to add a reference to the documentation.

Feel free to close this issue.

Cheers
Sebastian

**Comment by pmoulon** [comment]
Thank you @SBCV.
This is a great contribution?
Would you be ok to share a tiny demo here (some screenshot for the community).

- Does Blender is having a spherical camera model (if it is the case we could map Spherical SfM reconstruction in Blender...)?

**Comment by pmoulon** [comment]
@SBCV I made you a PR to make the import more robust to Views that have missing intrinsic or pose data (images that cannot be localized at the reconstruction stage).

**Comment by SBCV** [comment]
Sorry for the late reply.
Here is a screenshot of the visualization of a reconstruction of the [sceaux castle](https://github.com/openMVG/ImageDataset_SceauxCastle) in Blender 2.8. 

![import_result](https://user-images.githubusercontent.com/19491335/63195285-78fa8700-c072-11e9-90c4-d9a0c6be2369.jpg)

It is possible to add an image plane for each reconstructed camera pose. The point cloud is represented with an particle system. Also the addon allows to compute a camera animation (see image below).

![my3](https://user-images.githubusercontent.com/19491335/63196283-32f2f280-c075-11e9-8207-b53e4c03de84.gif)

Concerning the spherical camera model, I am not exactly sure what you are looking for. Also, I have no experience with such camera models in Blender.
There are a few options though (See image below).

![camera_model](https://user-images.githubusercontent.com/19491335/63196893-aea16f00-c076-11e9-9629-16ec2ad9b6f6.JPG)

And it is possible to render cameras with radial distortion.

https://docs.blender.org/manual/en/dev/compositing/types/distort/lens_distortion.html
I hope this gives a short overview of the addon and answers your questions (at least partially?)

**Comment by pmoulon** [comment]
Yes the equirectangular camera model would be the one corresponding to the 360 camera model of OpenMVG.
Great to know that Blender proposes this by!

- Let me know how we could work on adding support for those cameras.
I have some dataset and instruction there https://github.com/openMVG/Image_datasets/tree/master/360_dataset

PS: See also the PR I made regarding making the importer more robust.

**Comment by SBCV** [comment]
I'll have a look tomorrow.

The PR is already merged :) Thank you for the contribution!

**Comment by SBCV** [comment]
Can you share an example file?

**Comment by pmoulon** [comment]
Here you go [360_sfm_data.zip](https://github.com/openMVG/openMVG/files/3531013/360_sfm_data.zip), you have two sfm_data.json that correspond to https://github.com/openMVG/Image_datasets/tree/master/360_dataset

Thank you for your help.
I think we just have to create the instance of the equirectangular camera if we detect a spherical camera in the JSON file (see polymorphic_name `spherical`)
```
"intrinsics": [
        {
            "key": 0,
            "value": {
                "polymorphic_id": 2147483649,
                "polymorphic_name": "spherical",
                "ptr_wrapper": {
                    "id": 2147483660,
                    "data": {
                        "value0": {
                            "width": 5376,
                            "height": 2688
                        }
                    }
                }
            }
        }
    ]
```

**Comment by SBCV** [comment]
Thank you for sharing the dataset!
I've added support for spherical cameras in the last commit. 
When importing the cameras and the point cloud, make sure that the "Add background image for each camera" option is enabled. 
Then, select one of the imported cameras and press ctrl+0 (on the numpad) in order to view the point cloud from the perspective of the spherical camera. 
Alternatively, you can render the result. 
I hope this is the functionality you are looking for?
One small caveat: Viewing the points with a spherical camera causes (obviously) distortion of the points, since the addon represents the point cloud with a particle system. 
To reduce this effect make sure you adjust the size of the particles. [Here](https://github.com/SBCV/Blender-Addon-Photogrammetry-Importer/blob/master/doc/markdown/adjustment.md) is  short explanation.
